File: smart-backend/src/Smart.Api/Clients/zero-api-client.js
/**
 * ZERO APIクライアントのラッパー
 * 
 * ◆メモ
 * ・AUTH_KEYでのログインを実施して、cookieとしてfuelfidを取得。以後そのfuelfidをcookieとして送信して、ZERO APIを呼び出す。
 * ・実装を簡単にするため、一旦は各APIを呼び出す前に必ずauthを呼ぶことにする
 * 
 * ◆各APIの共通仕様
 * ・エラー時のレスポンス例（全API共通）:
 * {
 *   "ERROR_CODE": "6",
 *   "DATA": {
 *     "ERROR_DETAIL": [
 *       {
 *         "PARAM_NAME": "OWNER_ID",
 *         "ROW": ""
 *       }
 *     ]
 *   }
 * }
 */
class ZeroApiClient {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
        this.sessionString = null;
    }

    /**
     * 認証して、fuelfidを取得
     *
     * @param {*} appKey
     * @param {*} authKey
     * @param {*} ownerId
     * @param {*} areaId
     *
     * レスポンスのサンプル:
     * {
     *   "ERROR_CODE": "0",
     *   "DATA": {
     *     "LOGIN_INFO": [
     *       {
     *         "OWNER_ID": "1",
     *         "OWNER_DISP_NAME": "荷主01",
     *         "AREA_ID": "1",
     *         "AREA_DISP_NAME": "倉庫01",
     *         "LANG_ID": "JPN",
     *         "DEBUG_FLG": "0"
     *       }
     *     ]
     *   }
     * }
     */
    async auth(appKey, authKey, ownerId, areaId) {
        const loginResponse = await fetch(`${this.baseUrl}/login/login/keylogin`, {
            method: 'POST',
            body: JSON.stringify({
                APP_KEY: appKey,
                AUTH_KEY: authKey,
                OWNER_ID: ownerId,
                AREA_ID: areaId,
            }),
            headers: {
                'Content-Type': 'application/json'
            },
        });

        if (!loginResponse.ok) {
            throw new Error(`Login failed: ${loginResponse.statusText}`);
        }

        // Set-CookieヘッダーからCookieを取得
        const cookies = loginResponse.headers.get('set-cookie').split('; ');
        const fuelfid = cookies.filter(cookie => cookie.includes('fuelfid='))
            .map(cookie => cookie.match(/^.*(fuelfid=.*)$/)[1])
            .find(cookie => cookie !== 'fuelfid=deleted');
        this.sessionString = fuelfid;
        return await loginResponse.json();
    }

    /**
     * 梱包形態(重量)を更新
     *
     * @param {*} itemId 商品ID
     * @param {*} caseBarcode ケースバーコード
     * @param {*} caseWeight ケース_重量（SKU単位）
     */
    async updatePackageWeight(itemId, caseBarcode, { caseWeight = "" }) {
        const importResponse = await fetch(`${this.baseUrl}/common/import/import`, {
            method: 'POST',
            body: JSON.stringify({
                OWNER_ID: process.env.ZERO_API_OWNER_ID,
                AREA_ID: process.env.ZERO_API_AREA_ID,
                ONLY_AREA_IMPORT_FLG: "1",
                FILE_ID: "2115", // 商品マスタ(梱包形態)2
                PTRN_ID: "0", // スマートマット検証用パターン
                ERROR_DETAIL: "1",
                IMPORT_DATA: [
                    '"商品ID","ケースバーコード","ケース_重量（SKU単位）"',
                    `"${itemId}","${caseBarcode}","${caseWeight}"`,
                ]
            }),
            headers: {
                'Content-Type': 'application/json',
                'Cookie': this.sessionString
            }
        });
        if (!importResponse.ok) {
            throw new Error(`Update item package weight failed: ${importResponse.statusText}`);
        }
        return await importResponse.json();
    }

    /**
     * 梱包形態(サイズ)を更新
     *
     * @param {*} itemId 商品ID
     * @param {*} caseBarcode ケースバーコード
     * @param {*} caseLength ケース_縦（SKU単位）
     * @param {*} width ケース_横（SKU単位）
     * @param {*} height ケース_高さ（SKU単位）
     */
    async updatePackageSize(itemId, caseBarcode, { caseLength = "", caseWidth = "", caseHeight = "" }) {
        const importResponse = await fetch(`${this.baseUrl}/common/import/import`, {
            method: 'POST',
            body: JSON.stringify({
                OWNER_ID: process.env.ZERO_API_OWNER_ID,
                AREA_ID: process.env.ZERO_API_AREA_ID,
                ONLY_AREA_IMPORT_FLG: "1",
                FILE_ID: "2115", // 商品マスタ(梱包形態)2
                PTRN_ID: "1", // スマートマット&スマートメジャー検証用パターン
                ERROR_DETAIL: "1",
                IMPORT_DATA: [
                    '"商品ID","ケースバーコード","ケース_縦（SKU単位）","ケース_横（SKU単位）","ケース_高さ（SKU単位）"',
                    `"${itemId}","${caseBarcode}","${caseLength}","${caseWidth}","${caseHeight}"`,
                ]
            }),
            headers: {
                'Content-Type': 'application/json',
                'Cookie': this.sessionString
            }
        });

        if (!importResponse.ok) {
            throw new Error(`Update item package failed: ${importResponse.statusText}`);
        }
        return await importResponse.json();
    }
}

module.exports = ZeroApiClient;

File: smart-backend/src/Smart.Api/Clients/ZeroApiClient.cs
// ZeroApiClient.cs
// ZERO API 呼び出しの実装（zero-api-client.js 準拠: auth, updatePackageWeight, updatePackageSize, getSku）

using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Options;
using Smart.Api.Clients;
using Smart.Api.Options;
using Smart.Api.Extensions;

namespace Smart.Api.Clients;

/// <summary>
/// ZERO API を呼び出す HTTP クライアントの実装。
/// 認証で fuelfid を取得し、以降の import 呼び出しで Cookie を付与する。
/// </summary>
public class ZeroApiClient : IZeroApiClient
{
    private readonly HttpClient _httpClient;
    private readonly ZeroApiOptions _options;
    private string? _sessionCookie;

    /// <summary>
    /// コンストラクタ
    /// </summary>
    /// <param name="httpClient">HttpClient</param>
    /// <param name="options">ZERO API 設定</param>
    public ZeroApiClient(HttpClient httpClient, IOptions<ZeroApiOptions> options)
    {
        _httpClient = httpClient;
        _options = options.Value;
    }

    private string BaseUrl => _options.BaseUrl?.TrimEnd('/') ?? "";

    /// <inheritdoc />
    public async Task<object?> AuthAsync(CancellationToken cancellationToken = default)
    {
        var url = $"{BaseUrl}/login/login/keylogin";
        var body = JsonSerializer.Serialize(new
        {
            APP_KEY = _options.AppKey,
            AUTH_KEY = _options.AuthKey,
            OWNER_ID = _options.OwnerId,
            AREA_ID = _options.AreaId
        });
        using var content = new StringContent(body, Encoding.UTF8, "application/json");
        using var response = await _httpClient.PostAsync(url, content, cancellationToken);

        if (!response.IsSuccessStatusCode)
            throw new InvalidOperationException($"Login failed: {response.StatusCode}");

        _sessionCookie = ParseFuelfidFromSetCookie(response);
        var json = await response.Content.ReadAsStringAsync(cancellationToken);
        return string.IsNullOrWhiteSpace(json) ? null : JsonSerializer.Deserialize<JsonElement>(json);
    }

    /// <inheritdoc />
    public async Task<object?> GetSkuAsync(string pack_id, string barcode, CancellationToken cancellationToken = default)
    {
        await EnsureAuthenticatedAsync(cancellationToken);

        var url = $"{BaseUrl}/common/master/getsku";
        var body = JsonSerializer.Serialize(new
        {
            OWNER_ID = _options.OwnerId,
            PACK_ID = pack_id,
            BARCODE = barcode
        });

        using var request = new HttpRequestMessage(HttpMethod.Post, url);
        request.Content = new StringContent(body, Encoding.UTF8, "application/json");
        var response = await _httpClient.SendAsync(request, cancellationToken);

        if (!response.IsSuccessStatusCode)
            return null;

        var json = await response.Content.ReadAsStringAsync(cancellationToken);
        if (string.IsNullOrWhiteSpace(json))
            return null;

        return JsonSerializer.Deserialize<JsonElement>(json);
    }

    public async Task<object?> ImportAsync(ImportRequest importRequest, CancellationToken cancellationToken = default)
    {
        await EnsureAuthenticatedAsync(cancellationToken);

        Console.WriteLine(string.Join("\n", importRequest.ImportCsvRows));

        var body = new
        {
            OWNER_ID = _options.OwnerId,
            AREA_ID = _options.AreaId,
            ONLY_AREA_IMPORT_FLG = "1",
            FILE_ID = importRequest.FileId,
            PTRN_ID = importRequest.PtrnId,
            ERROR_DETAIL = "1",
            IMPORT_DATA = importRequest.ImportCsvRows
        };

        var url = $"{BaseUrl}/common/import/import";
        using var request = new HttpRequestMessage(HttpMethod.Post, url);
        request.Headers.Add("Cookie", _sessionCookie);
        request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
            throw new InvalidOperationException($"Import failed: {response.StatusCode}");

        var json = await response.Content.ReadAsStringAsync(cancellationToken);
        return string.IsNullOrWhiteSpace(json) ? null : JsonSerializer.Deserialize<JsonElement>(json);
    }

    private async Task EnsureAuthenticatedAsync(CancellationToken cancellationToken)
    {
        if (string.IsNullOrEmpty(_sessionCookie))
            await AuthAsync(cancellationToken);
    }

    /// <summary>
    /// Set-Cookie ヘッダから fuelfid を取得（Cookie ヘッダ用に "fuelfid=値" を返す）
    /// </summary>
    private static string? ParseFuelfidFromSetCookie(HttpResponseMessage response)
    {
        if (!response.Headers.TryGetValues("Set-Cookie", out var values))
            return null;

        foreach (var header in values)
        {
            var parts = header.Split(';');
            var first = parts[0].Trim();
            if (first.StartsWith("fuelfid=", StringComparison.OrdinalIgnoreCase) &&
                !string.Equals(first, "fuelfid=deleted", StringComparison.OrdinalIgnoreCase))
                return first;
        }

        return null;
    }

}
